### PR TITLE
TDL-18217 change lookback window logic

### PR DESCRIPTION
**Pull request by namrata270998**
# Description of change
- Added a configurable lookback window for both the balance transactions as well as the events stream
- Changed the start time evaluation logic for both the streams

# Manual QA steps
 - Added unittests to ensure both the streams captures correct data with lookback window
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/test_configurable_lookback_window.py
Comment (savan-chovatiya) [inline tests/test_configurable_lookback_window.py:31]: @namrata270998 Can you please correct the spell of client_secret here because the test is failing for this reason? Also, this test doesn't have any assert statement so what are we checking here?

File: tap_stripe/__init__.py
Comment (savan-chovatiya) [inline tap_stripe/__init__.py:434]: Add function comment here.

File: tests/unittests/test_lookback_window.py
Comment (savan-chovatiya) [inline tests/unittests/test_lookback_window.py:56]:
```suggestion
        # expected start_date should be the now() - `lookback`(default lookback)
```

File: tests/test_all_fields.py
Comment (namrata270998) [reply tests/test_all_fields.py]: Move the logic to separate func

File: tests/unittests/test_lookback_window.py
Comment (savan-chovatiya) [inline tests/unittests/test_lookback_window.py:86]:
```suggestion
        # expected start_date should be the now() - `lookback`(default lookback)
```

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tests/test_configurable_lookback_window.py
Comment (namrata270998) [reply tests/test_configurable_lookback_window.py]: This test was suggested by Kyle on the other PR for balance transactions lookback [here](https://github.com/singer-io/tap-stripe/pull/125#pullrequestreview-900330370). It is just to check that the lookback window is added as a UI parameter or not

File: tests/unittests/test_lookback_window.py
Comment (namrata270998) [reply tests/unittests/test_lookback_window.py]: Updated

File: tap_stripe/__init__.py
Comment (namrata270998) [reply tap_stripe/__init__.py:434]: Added comment

File: tests/unittests/test_lookback_window.py
Comment (namrata270998) [reply tests/unittests/test_lookback_window.py]: Updated

**Review by umangagrawal-crest — COMMENTED** [review]
Add the newly added parameter in the sample config file and README

File: tap_stripe/__init__.py
Comment (umangagrawal-crest) [inline tap_stripe/__init__.py:545]:
```suggestion
            if lookback_window and float(lookback_window): # set lookback window if lookback window is present in config and float convertible
                lookback_window = float(lookback_window)
```
```suggestion
            if lookback_window and int(lookback_window): # set lookback window if lookback window is present in config and int convertible
                lookback_window = int(lookback_window)
```

Comment (namrata270998) [reply tap_stripe/__init__.py]: Updated to accept float values

**Review by savan-chovatiya — APPROVED** [review]

**Review by dbshah1212 — APPROVED** [review]

File: tests/test_configurable_lookback_window.py
Comment (kspeer825) [inline tests/test_configurable_lookback_window.py:53]: copy pasta, should be removed

File: tests/test_configurable_lookback_window.py
Comment (kspeer825) [inline tests/test_configurable_lookback_window.py:60]: Why is the variable declared, then immediately overwritten?

File: tests/test_configurable_lookback_window.py
Comment (kspeer825) [inline tests/test_configurable_lookback_window.py:67]: If this is only setting state for 1 stream, it doesn't need to be a dictionary comprehension like this.

File: .circleci/config.yml
Comment (kspeer825) [inline .circleci/config.yml:219]: The test_lookback_window.py file does not exist.

**Review by kspeer825 — CHANGES_REQUESTED** [review]

File: tests/test_configurable_lookback_window.py
Comment (namrata270998) [reply tests/test_configurable_lookback_window.py]: Removed the unused var

File: .circleci/config.yml
Comment (namrata270998) [reply .circleci/config.yml]: This file was added before and removed due to a change in bookmark logic but left to remove in the config. Removed it

File: tests/test_configurable_lookback_window.py
Comment (namrata270998) [reply tests/test_configurable_lookback_window.py]: Removed the unwanted initialization

**Review by KrisPersonal — APPROVED** [review]

**Review by RushT007 — CHANGES_REQUESTED** [review]

File: tests/test_all_fields.py
Comment (kspeer825) [inline tests/test_all_fields.py:584]: This pattern should be removed, see PR 137

File: tests/test_configurable_lookback_window.py
Comment (kspeer825) [inline tests/test_configurable_lookback_window.py:28]: I don't think the 600 seconds is doing anything. The start date format should not include time infomation only date. So this should evaluate to something like `05-25-22T00:00:00Z`.

File: tests/test_configurable_lookback_window.py
Comment (kspeer825) [inline tests/test_configurable_lookback_window.py:65]: There are duplicate declarations for  initial_state. 
Also I am not sure that a timestamp should be a valid bookmark format for a datetime value. Not really a functional issue, and if that is the existing behavior that I don't think it's worth altering the tap, but for consistency across taps in the future I think we prefer bookmarked datetimes to be formatted as ISO 8601/RFC 3339 utc.

**Review by kspeer825 — CHANGES_REQUESTED** [review]

File: tests/test_configurable_lookback_window.py
Comment (namrata270998) [reply tests/test_configurable_lookback_window.py]: - Removed the duplicate declaration. 
- Stripe uses timestamp as a format for the bookmark value.
- Sure, we'll take care of this in the future, if a tap has bookmark as timestamp we'll convert it to ISO 8601/RFC 3339 utc.

File: tests/test_configurable_lookback_window.py
Comment (namrata270998) [reply tests/test_configurable_lookback_window.py:28]: The `START_DATE_FORMAT` was already defined and used in the `test_start_date.py`, hence we've utilized the same. Moreover, the `timedelta(600)` is just to create spare time between the start date and the sync time, hence we can get lookback.

File: tests/unittests/test_lookback_evaluation.py
Comment (namrata270998) [reply tests/unittests/test_lookback_evaluation.py]: Added comments in the unittests

File: tests/test_configurable_lookback_window.py
Comment (kspeer825) [reply tests/test_configurable_lookback_window.py:28]: I see, but  subtracting 10 minutest from today's date is unlikely to actually change the value that you set as start_date. It's going to be utc midnight of today. Applying the start date format https://github.com/singer-io/tap-stripe/blob/cbdd805335adc6031250cad6719ed2f2c82e8be7/tests/base.py#L29 will truncate the time information from the datetime value. Not a big deal, but just so you are aware.

**Review by kspeer825 — APPROVED** [review]
Responded to a comment, but functionally the tests look good 👍

File: tests/test_all_fields.py
Comment (namrata270998) [reply tests/test_all_fields.py]: Removed the function itself

File: tests/test_configurable_lookback_window.py
Comment (namrata270998) [reply tests/test_configurable_lookback_window.py:28]: Ohh understood what you meant, the time information would be made 0. Thanks for explaining.

**Review by dbshah1212 — APPROVED** [review]

File: tests/test_configurable_lookback_window.py
Comment (RushT007) [reply tests/test_configurable_lookback_window.py:28]: Lets remove `timedelta(seconds=600)` if it doesn't have any significance.

**Review by RushT007 — CHANGES_REQUESTED** [review]

**Comment by RushT007** [comment]
Currently we are waiting for other PR to merge which will add lookback_window property to tap-stripe. Once that is done and CCi is passed I will approve this PR.

**Review by RushT007 — CHANGES_REQUESTED** [review]

File: .circleci/config.yml
Comment (namrata270998) [reply .circleci/config.yml:222]: Updated the dependency

File: tests/unittests/test_lookback_window.py
Comment (RushT007) [inline tests/unittests/test_lookback_window.py:88]: Empty string '' is an invalid value so this test should fail. Here you may need to change the assertion accordingly.

File: tests/test_configurable_lookback_window.py
Comment (namrata270998) [reply tests/test_configurable_lookback_window.py]: As per the discussion with @KrisPersonal, the current implementation makes sense. Hence moving forward with the same

File: tests/unittests/test_lookback_window.py
Comment (namrata270998) [reply tests/unittests/test_lookback_window.py:88]: As per the discussion with @KrisPersonal, the current implementation makes sense. Hence moving forward with the same

**Pull request merged**
